scripts/portal-fetcher/openshift-docs-downloader.py

```diff
@@ -0,0 +1,1186 @@
+#!/usr/bin/env python3
+"""
+OpenShift Documentation Downloader - Functional Programming Version
+
+Downloads the HTML-single version of Red Hat OpenShift documentation for a specified version,
+preserving directory structure and maintaining a database of downloaded files.
+Supports incremental updates and advanced verification.
+
+This version is designed with functional programming principles:
+- Pure functions that avoid side effects where possible
+- Immutable data structures
+- Function composition
+- Higher-order functions
+"""
+
+import argparse
+import asyncio
+import aiohttp
+import os
+import json
+import functools
+import logging
+import time
+import datetime
+from urllib.parse import urljoin, urlparse
+from bs4 import BeautifulSoup
+from pathlib import Path
+import sqlite3
+
+# Configure logging
+logging.basicConfig(
+    level=logging.INFO,
+    format='%(asctime)s - %(levelname)s - %(message)s',
+    handlers=[
+        logging.FileHandler("openshift-downloader.log"),
```

Comment:
Logging to a hardcoded file is against https://12factor.net/logs. It's enough to log to the console. If need be, the log will be tee'd or redirected to a file.